lib-ltc/src/ltc.cpp

```diff
@@ -256,3 +256,142 @@ bool Ltc::ParseTimeCodeRate(const char *pTimeCodeRate, uint8_t &nFPS, type &tTyp
+		return __UINT32_MAX__; // 4294967295
+		break;
+	}
+
+	double fps_d = div;
+	double fps_i = _tc_round_up(div);
+	double fpf_d = (1000 / fps_d);
+	double sample = 0;
+	if (drop)
+	{
+		double tMin = 60 * ptLtcTimeCode->nHours + ptLtcTimeCode->nMinutes;
+		double frameNumber = fps_i * 3600 * ptLtcTimeCode->nHours + fps_i * 60 * ptLtcTimeCode->nMinutes + fps_i * ptLtcTimeCode->nSeconds + ptLtcTimeCode->nFrames - 2 * (tMin - tMin / 10);
+		sample = frameNumber * 1000 / fps_d;
+	}
+	else
+	{
+		sample = _tc_round((((ptLtcTimeCode->nHours * 60 * 60) + (ptLtcTimeCode->nMinutes * 60) + ptLtcTimeCode->nSeconds) * (fps_i * fpf_d)) + (ptLtcTimeCode->nFrames * fpf_d));
+	}
+	return static_cast<uint32_t>(sample);
+}
+
+// take a value in ms and set ptLtcTimeCode h,m,s,f based on TC type already in ptLtcTimeCode->nType
+// return non-zero on fail
+uint32_t Ltc::MStoTC(const uint32_t ms, struct TLtcTimeCode *ptLtcTimeCode) {
+	bool drop = false;
+	double div = 0.0;
+
+	switch (ptLtcTimeCode->nType) {
+	case type::FILM:
+		div = 24.0;
+		break;
+	case type::EBU:
+		div = 25.0;
+		break;
+	case type::DF:
+		div = 30.0;
+		drop = true;
+		break;
+	case type::SMPTE:
+		div = 30.0;
+		break;
+	default:
+		return __UINT32_MAX__; // 4294967295
+		break;
+	}
+
+	uint32_t fps_i = _tc_round_up(div);
+	if (fps_i == 0)	{
+		return 1; // invalid TC rate
+	}
+
+	if (ms == 0) {
+		ptLtcTimeCode->nHours = 0;
+		ptLtcTimeCode->nMinutes = 0;
+		ptLtcTimeCode->nSeconds = 0;
+		ptLtcTimeCode->nFrames = 0;
+		return 1;
+	}
+
+	if (drop) {
+		uint32_t frameNumber = _tc_round_down(ms * div / 1000);
+		uint32_t D = frameNumber / 17982; // there are 17982 frames in 10 min @ 29.97df
+		uint32_t M = frameNumber % 17982;
+
+		frameNumber += (18 * (D + (2 * ((M - 2) / 1798))));
+		ptLtcTimeCode->nFrames = frameNumber % 30;
+		ptLtcTimeCode->nSeconds = (frameNumber / 30) % 60;
+		ptLtcTimeCode->nMinutes = ((frameNumber / 30) / 60) % 60;
+		ptLtcTimeCode->nHours = (((frameNumber / 30) / 60) / 60);
+	}
+	else {
+		static double timecode_frames_left_exact;
+		static uint32_t timecode_frames_left;
+		double fpf_d = 1000 / div;
+		double frames_per_hour = 3600 * fps_i * fpf_d;
+		double sample_d = (ms % static_cast<uint32_t>(frames_per_hour));
+
+		timecode_frames_left_exact = sample_d / fpf_d;
+		timecode_frames_left = _tc_round(timecode_frames_left_exact);
+		ptLtcTimeCode->nHours = _tc_round(ms / frames_per_hour);
+		ptLtcTimeCode->nMinutes = timecode_frames_left / (fps_i * 60);
+		if (ptLtcTimeCode->nMinutes == 60) {
+			ptLtcTimeCode->nMinutes = 0;
+		}
+		timecode_frames_left = timecode_frames_left % (fps_i * 60);
+		ptLtcTimeCode->nSeconds = static_cast<uint8_t>(timecode_frames_left / fps_i);
+		ptLtcTimeCode->nFrames = static_cast<uint8_t>(timecode_frames_left % fps_i);
+	}
+
+	if (ptLtcTimeCode->nHours > 23) { // don't allow >= 24hrs
+		ptLtcTimeCode->nHours = 0;
+		ptLtcTimeCode->nMinutes = 0;
+		ptLtcTimeCode->nSeconds = 0;
+		ptLtcTimeCode->nFrames = 0;
+	}
+
+	return 0;
+}
```

Comment:
ms to TimeCode is already here -> https://github.com/vanvught/rpidmx512/blob/master/lib-tcnet/src/tcnet.cpp#L128
Implementation should be similar.

Reply:
Thanks, I ported my own code without realizing it was available, I will adapt.